modules/imgcodecs/src/grfmt_qoi.cpp

```diff
@@ -0,0 +1,297 @@
+//
+// This software is provided by the copyright holders and contributors "as is" and
+// any express or implied warranties, including, but not limited to, the implied
+// warranties of merchantability and fitness for a particular purpose are disclaimed.
+// In no event shall the Intel Corporation or contributors be liable for any direct,
+// indirect, incidental, special, exemplary, or consequential damages
+// (including, but not limited to, procurement of substitute goods or services;
+// loss of use, data, or profits; or business interruption) however caused
+// and on any theory of liability, whether in contract, strict liability,
+// or tort (including negligence or otherwise) arising in any way out of
+// the use of this software, even if advised of the possibility of such damage.
+//
+//M*/
+
+#include "precomp.hpp"
+#include "utils.hpp"
+#include "bitstrm.hpp"
+#include "grfmt_qoi.hpp"
+#include <opencv2/core/utils/logger.hpp>
+
+#ifdef HAVE_QOI
+
+#define QOI_IMPLEMENTATION
+#include "qoi.h"
+
+namespace cv
+{
+
+QoiDecoder::QoiDecoder()
+{
+    m_signature = "qoif";
+    m_buf_supported = true;
+}
+
+QoiDecoder::~QoiDecoder()
+{
+}
+
+static void swap_endianess(int32_t& val)
+{
+    static const uint32_t A(0x000000ffU);
+    static const uint32_t B(0x0000ff00U);
+    static const uint32_t C(0x00ff0000U);
+    static const uint32_t D(0xff000000U);
+
+    union {
+        uint32_t asU32;
+        int32_t  asS32;
+    } tmp;
+
+    tmp.asS32 = val;
+
+    tmp.asU32 = ( (tmp.asU32 & A) << 24 )
+              | ( (tmp.asU32 & B) <<  8 )
+              | ( (tmp.asU32 & C) >>  8 )
+              | ( (tmp.asU32 & D) >> 24 );
+
+    val = tmp.asS32;
+}
+
+bool QoiDecoder::readHeader()
+{
+    RLByteStream strm;
+
+    if( !m_buf.empty() )
+    {
+        CV_Assert( m_buf.type() == CV_8UC1 );
+        CV_Assert( m_buf.rows == 1 );
+
+        size_t mbuf_len = m_buf.cols * m_buf.rows * m_buf.elemSize();
+        if( mbuf_len < QOI_HEADER_SIZE + sizeof(qoi_padding) )
+        {
+            CV_LOG_WARNING( NULL, "QOI: readHeader(): Too short data." );
+            return false;
+        }
+        uint8_t* padding_pos = m_buf.data + mbuf_len - sizeof(qoi_padding);
+        if( memcmp( padding_pos, qoi_padding, sizeof(qoi_padding) ) != 0 )
+        {
+            CV_LOG_WARNING( NULL, "QOI: readHeader(): missing padding data." );
+            return false;
+        }
+        if( !strm.open(m_buf) )
+        {
+            return false;
+        }
+    }
+    else if( !strm.open( m_filename ) )
+    {
+        return false;
+    }
+
+    strm.skip(4); // Signature
+    m_width    = strm.getDWord(); // As Big Endian
+    m_height   = strm.getDWord(); // As Bit Endian
+    int ncn    = strm.getByte();
+
+    if( !isBigEndian() )
+    {
+        swap_endianess( m_width );  // As Little Endian
+        swap_endianess( m_height ); // As Little Endian
+    }
+
+    CV_Assert( ncn == 3 || ncn == 4 );
```

Comment:
`CV_Check*` macro is preferable.

`CV_Check(ncn, ncn == 3 || ncn == 4, "");`

Reply:
I believe all CV_Assert() macros has been replaced to CV_Check*().
I have omitted some comments(3rd argument) that are self-explanatory.